File: oblig2/Simpsons.py
import sys
from rdflib import Graph, Namespace, Literal, BNode


def main():

    # Command line arguments
    if len(sys.argv) != 3:
        print(f"Missing arguments <input_file> <output_file>")

    input_path = sys.argv[1]
    output_path = sys.argv[2]

    graph = Graph()

    # Task 1
    try:
        graph.parse(input_path, format="turtle")
    except Exception as e:
        print(f"error parsing ttl file: {e}")

    # read through namespaces, add to graph
    for prefix, namespace in graph.namespaces():
        ns = Namespace(namespace)
        graph.bind(prefix, ns)

        print(f"@Prefix {prefix}: <{namespace}>")

    # Task 2
    manager = graph.namespace_manager.store
    SIM = Namespace(manager.namespace("sim"))
    FAM = Namespace(manager.namespace("fam"))
    RDF = Namespace(manager.namespace("rdf"))
    FOAF = Namespace(manager.namespace("foaf"))
    XSD = Namespace(manager.namespace("xsd"))

    # Maggie is a person, name Maggie Simpsons, age 1 of datatype xsd:int
    maggie = SIM.Maggie  # reference to Maggie
    graph.add((maggie, RDF.type, FOAF.Person))
    graph.add((maggie, FOAF.name, Literal("Maggie Simpson")))
    graph.add((maggie, FOAF.age, Literal(1, datatype=XSD.int)))

    # Mona is a person, name Mona Simpson, age 70 of datatype xsd:int
    mona = SIM.Mona     # reference to Mona
    graph.add((mona, RDF.type, FOAF.Person))
    graph.add((mona, FOAF.name, Literal("Mona Simpson")))
    graph.add((mona, FOAF.age, Literal(70, datatype=XSD.int)))

    # Abraham is a person, name Abraham Simpson, age  78 of datatype xsd:int
    abraham = SIM.Abraham     # reference to Abraham
    graph.add((mona, RDF.type, FOAF.Person))
    graph.add((mona, FOAF.name, Literal("Abraham Simpson")))
    graph.add((mona, FOAF.age, Literal(78, datatype=XSD.int)))

    # Abraham is spouse of Mona, Mona is spouse of Abraham
    graph.add((abraham, FAM.hasSpouse, mona))
    graph.add((mona, FAM.hasSpouse, abraham))

    # Herb is a person, has an unknown father
    herb = SIM.Herb     # reference to Herb
    father = BNode()    # reference to blank father
    graph.add((herb, RDF.type, FOAF.Person))
    graph.add((herb, FAM.hasFather, father))

    # Task 3
    # printer hvert subjekt
    print("\n\nPersoner:")
    for person in graph.subjects(unique=True):
        print(person)

        #TODO:
        # filtrer subjekt på FOAF:Person
        # hent alder
        # sjekk alder < 2 = infant, < 18 = Minor, > 70 = Old
        # legg til   .add((SIM.navn, RDF.type, FAM.Old/Minor/Infant))


    # Task 4
    with open(output_path, "w") as f:
        f.write(graph.serialize(format="turtle"))


if __name__ == "__main__":
    main()

File: oblig2/java/Simpsons.java
import org.apache.jena.datatypes.xsd.XSDDatatype;
import org.apache.jena.rdf.model.ModelFactory;
import org.apache.jena.rdf.model.Statement;
import org.apache.jena.rdf.model.Resource;
import org.apache.jena.rdf.model.Property;
import org.apache.jena.rdf.model.Literal;
import org.apache.jena.rdf.model.Model;

import java.io.IOException;
import java.io.FileReader;
import java.io.FileWriter;
import java.util.Iterator;
import java.io.Reader;
import java.io.Writer;
import java.util.Map;
import java.io.File;


public class Simpsons {

    // I/O
    static String input_path;
    static String output_path;

    // Global namespaces
    static String SIM;
    static String FAM;
    static String RDF;
    static String FOAF;
    static String XSD;


    /**
     * Takes in a string and determines file-format for jena usage
     * @param url 
     * @return format for jena
     */
    private static String fileFormat(String url) {
        if (url.endsWith(".ttl")) {
            return "TURTLE";
        }
        else if (url.endsWith(".rdf")) {
            return "RDF/XML";
        }
        else if (url.endsWith(".n3")) {
            return "N3";
        }
        else if (url.endsWith(".nt")) {
            return "N-TRIPLES";
        }
        else return null;
    }


    public static void main(String[] args) {

        //TODO: Remove
        System.out.println("\n\n\n\n\n\n\n\n");

        if (args.length < 2) {
            System.out.println("Missing arguments <input_file> <output_file>");
            return;
        }

        input_path = args[0];
        output_path = args[1];

        
        /** Task 1 */

        // Creates model hold .ttl graph
        Model model = ModelFactory.createDefaultModel();

        // Read from input file to model
        try (Reader reader = new FileReader(new File(input_path))) {
            
            String format = fileFormat(input_path);
            model.read(reader, null, format);
        } 
        catch (IOException e) {
            e.printStackTrace();
        }

        System.out.println("Input Path: " + input_path);
        System.out.println("Output Path: " + output_path + "\n\n");
        
        // add namespaces to global scope
        Map<String, String> namespaces = model.getNsPrefixMap();
        SIM = namespaces.get("sim");
        FAM = namespaces.get("fam");
        RDF = namespaces.get("rdf");
        FOAF = namespaces.get("foaf");
        XSD = namespaces.get("xsd");
        

        /** Task 2 */
        Property hasSpouse = model.createProperty(FAM + "hasSpouse");
        Property hasFather = model.createProperty(FAM + "hasFather");
        Property foafName = model.createProperty(FOAF + "name");
        Property rdfType = model.createProperty(RDF + "type");
        Property foafAge = model.createProperty(FOAF + "age");
        Resource foafPerson = model.createResource(FOAF + "Person");

        // Maggie is a person named "Maggie Simpson" of age 1 datatype xsd:int
        Resource maggie = model.createResource(SIM + "Maggie")
            .addProperty(rdfType, foafPerson)
            .addProperty(foafAge, model.createTypedLiteral("1", XSDDatatype.XSDint))
            .addProperty(foafName, "Maggie Simpson");

        // Mona is a person named "Mona Simpson" of age 70 datatype xsd:int
        Resource mona = model.createResource(SIM + "Mona")
            .addProperty(rdfType, foafPerson)
            .addProperty(foafAge, model.createTypedLiteral("70", XSDDatatype.XSDint))
            .addProperty(foafName, "Mona Simpson");
       
        // Abraham is a person named "Abraham Simpson" of age 78 datatype xsd:int
        Resource abraham = model.createResource(SIM + "Abraham")
            .addProperty(rdfType, foafPerson)
            .addProperty(foafAge, model.createTypedLiteral("78", XSDDatatype.XSDint))
            .addProperty(foafName, "Abraham Simpson");

        // Abraham is spouse to Mona and vice versa
        model.add(abraham, hasSpouse, mona);
        model.add(mona, hasSpouse, abraham);

        // Herb is a person, and has _a_ father
        Resource blank = model.createResource();
        Resource herb = model.createResource(SIM + "Herb")
            .addProperty(rdfType, foafPerson)
            .addProperty(hasFather, blank);


        /** Task 3 */
        Iterator<Resource> it = model.listSubjectsWithProperty(rdfType, foafPerson);

        while (it.hasNext()) {
            Resource person = it.next();
            
            Statement ageStmt = person.getProperty(foafAge);
            if (ageStmt != null) {
                int age = ageStmt.getInt();

                // check age, add age-type
                if (age < 2) {
                    model.add(person, rdfType, model.createResource(FAM + "Infant"));
                }
                if (age < 18) {
                    model.add(person, rdfType, model.createResource(FAM + "Minor"));
                }
                if (age > 70) {
                    model.add(person, rdfType, model.createResource(FAM + "Old"));
                }
            }
        }


        /** Task 4 */
        try (Writer writer = new FileWriter(new File(output_path))) {
            
            String format = fileFormat(output_path);
            model.write(writer, format);

            model.write(System.out, format); // DEBUG
        } 
        catch (IOException e) {
            e.printStackTrace();
        }
    }
}
